File: src/org/simbrain/network/gui/nodes/SynapseNode.java
/*
 * Part of Simbrain--a java-based neural network kit
 * Copyright (C) 2005,2007 The Authors.  See http://www.simbrain.net/credits
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
 */
package org.simbrain.network.gui.nodes;

import org.piccolo2d.nodes.PPath;
import org.piccolo2d.util.PBounds;
import org.simbrain.network.core.Synapse;
import org.simbrain.network.gui.NetworkPanel;
import org.simbrain.network.gui.dialogs.synapse.SynapseDialog;

import javax.swing.*;
import java.awt.*;
import java.awt.geom.Arc2D;
import java.awt.geom.Line2D;
import java.awt.geom.Point2D;

/**
 * <b>SynapseNode</b> is a Piccolo PNode corresponding to a Neuron in the neural
 * network model.
 */
public final class SynapseNode extends ScreenElement {

    /**
     * The logical synapse this screen element represents.
     */
    private Synapse synapse;

    /**
     * Location of circle relative to target node.
     */
    private final double offset = 7;

    /**
     * Main circle of synapse.
     */
    private PPath circle;

    /**
     * Line connecting nodes. More of a loop for self cocnections.
     */
    private PPath.Float line;

    /**
     * The line for bound checking.
     */
    private Line2D.Float lineBound = new Line2D.Float();

    /**
     * The arc for bound checking when the synapse is a self connection.
     */
    private Arc2D.Float arcBound = new Arc2D.Float();

    /**
     * Reference to source neuron.
     */
    private NeuronNode source;

    /**
     * Reference to target neuron.
     */
    private NeuronNode target;

    /**
     * Used to approximate zero to prevent divide-by-zero errors.
     */
    private static final double ZERO_PROXY = .001;

    /**
     * Color of "excitatory" synapses, with positive values.
     */
    private static Color excitatoryColor = Color.red;

    /**
     * Color of "inhibitory" synapses, with negative values.
     */
    private static Color inhibitoryColor = Color.blue;

    /**
     * Color of "zero" weights.
     */
    private static Color zeroWeightColor = Color.gray;

    /**
     * Maximum diameter of the circle representing the synapse.
     */
    private static int maxDiameter = 20;

    /**
     * Minimum diameter of the circle representing the synapse.
     */
    private static int minDiameter = 7;

    /**
     * Color of lines in synapse representation.
     */
    private static Color lineColor = Color.black;

    /**
     * Create a new synapse node connecting a source and target neuron.
     *
     * @param net     Reference to NetworkPanel
     * @param source  source neuronnode
     * @param target  target neuronmode
     * @param synapse the model synapse this PNode represents
     */
    public SynapseNode(final NetworkPanel net, final NeuronNode source, final NeuronNode target, final Synapse synapse) {

        super(net);
        this.source = source;
        this.target = target;
        target.getConnectedSynapses().add(this);
        source.getConnectedSynapses().add(this);

        this.synapse = synapse;
        init();
    }

    /**
     * Initialize the SynapseNode.
     */
    private void init() {
        updatePosition();
        this.addChild(circle);
        this.addChild(line);
        line.setStrokePaint(Color.BLACK);
        line.lowerToBottom();

        updateColor();
        updateDiameter();

        setPickable(true);
        circle.setPickable(true);
        line.setPickable(false);
    }

    /**
     * Update position of synapse.
     */
    public void updatePosition() {

        Point2D synapseCenter;

        // Position the synapse
        if (isSelfConnection()) {
            synapseCenter = globalToLocal(new Point2D.Double(target.getCenter().getX() + offset + 3, target.getCenter().getY() + offset + 3));
        } else {
            synapseCenter = globalToLocal(calcCenter(source.getCenter(), target.getCenter()));
        }
        this.offset(synapseCenter.getX() - offset, synapseCenter.getY() - offset);

        // Create the circle
        if (circle == null) {
            circle = PPath.createEllipse(0, 0, (float) offset * 2, (float) offset * 2);
            ((PPath) circle).setStrokePaint(null);
        }
        double ea = 1;
        double ea2 = ea * 2;
        PBounds tempBounds = circle.getFullBounds();
        setBounds(tempBounds.x - ea, tempBounds.y - ea, tempBounds.width + ea2, tempBounds.height + ea2);

        // Create the line
        if (line == null) {
            line = createLine(globalToLocal(synapseCenter));
        }

        // Update the line (unless it's a self connection)
        if (!isSelfConnection()) {
            line.reset();
            line.append(new Line2D.Double(globalToLocal(source.getCenter()), synapseCenter), false);
            lineBound.setLine(source.getCenter(), localToGlobal(synapseCenter));
        } else {
            arcBound =
                    new Arc2D.Float(
                            (float) getGlobalBounds().getX(),
                            (float) getGlobalBounds().getY() - 7,
                            22,
                            15,
                            1,
                            355,
                            Arc2D.OPEN
                    );
        }
    }

    /**
     * of a synapse based on whether it is clamped or not.
     */
    public void updateClampStatus() {
        if (synapse.isFrozen()) {
            circle.setStrokePaint(Color.black);
        } else {
            circle.setStrokePaint(null);
        }
    }

    /**
     * Whether this synapse connects a neuron to itself or not.
     *
     * @return true if this synapse connects a neuron to itself.
     */
    public boolean isSelfConnection() {
        return (source.getNeuron() == target.getNeuron());
    }

    /**
     * Create the line depending on whether this is self connected or not.
     *
     * @param center the center of the synapse
     * @return the line
     */
    private PPath.Float createLine(final Point2D center) {
        if (isSelfConnection()) {
            return new PPath.Float(new Arc2D.Float((float) getX(), (float) getY() - 7, 22, 15, 1, 355, Arc2D.OPEN));
        } else {
            return new PPath.Float(new Line2D.Float(globalToLocal(source.getCenter()), center));
        }
    }

    /**
     * Calculates the color for a weight, based on its current strength.
     * Positive values are (for example) red, negative values blue.
     */
    public void updateColor() {
        if (synapse.getStrength() < 0) {
            circle.setPaint(inhibitoryColor);
        } else if (synapse.getStrength() == 0) {
            circle.setPaint(zeroWeightColor);
        } else {
            circle.setPaint(excitatoryColor);
        }
        if (source.getNeuron().isSpike()) {
            line.setStrokePaint(NeuronNode.getSpikingColor());
        } else {
            line.setStrokePaint(lineColor);
        }
    }

    /**
     * Update the diameter of the drawn weight based on the logical weight's
     * strength.
     */
    public void updateDiameter() {
        double diameter;

        double upperBound = synapse.getUpperBound();
        double lowerBound = synapse.getLowerBound();
        double strength = synapse.getStrength();

        // If upper or lower bound are set to zero use a proxy to prevent
        // division errors
        if (upperBound == 0) {
            upperBound = ZERO_PROXY;
        }
        if (lowerBound == 0) {
            lowerBound = ZERO_PROXY;
        }

        // If strength is out of bounds (which is allowed in the model), set it
        // to those bounds for the
        // sake of the GUI representation
        if (strength < lowerBound) {
            strength = lowerBound;
        }
        if (strength > upperBound) {
            strength = upperBound;
        }

        if (synapse.getStrength() == 0) {
            diameter = minDiameter;
        } else if (synapse.getStrength() > 0) {
            diameter = ((maxDiameter - minDiameter) * (strength / upperBound) + minDiameter);
        } else {
            diameter = (((maxDiameter - minDiameter) * (Math.abs(strength / lowerBound))) + minDiameter);
        }

        double delta = (circle.getBounds().getWidth() - diameter) / 2;

        circle.setWidth(diameter);
        circle.setHeight(diameter);
        // offset properly moves circle, but this is not reflected in bounds
        circle.offset(delta, delta);
        setBounds(circle.getFullBounds());
    }

    /**
     * Calculates the position of the synapse circle based on the positions of
     * the source and target NeuronNodes.
     *
     * @param src Source NeuronNode
     * @param tar Target NeuronNode
     * @return the appropriate position for the synapse circle
     */
    public Point2D calcCenter(final Point2D src, final Point2D tar) {

        double sourceX = src.getX();
        double sourceY = src.getY();
        double targetX = tar.getX();
        double targetY = tar.getY();

        double x = Math.abs(sourceX - targetX);
        double y = Math.abs(sourceY - targetY);
        double alpha = Math.atan(y / x);

        double weightX = 0;
        double weightY = 0;

        int neuronOffset = NeuronNode.getDIAMETER() / 2;

        if (sourceX < targetX) {
            weightX = targetX - (neuronOffset * Math.cos(alpha));
        } else {
            weightX = targetX + (neuronOffset * Math.cos(alpha));
        }

        if (sourceY < targetY) {
            weightY = targetY - (neuronOffset * Math.sin(alpha));
        } else {
            weightY = targetY + (neuronOffset * Math.sin(alpha));
        }

        return new Point2D.Double(weightX, weightY);
    }

    /**
     * @return
     * @see ScreenElement
     */
    public boolean isSelectable() {
        return true;
    }

    /**
     * @return
     * @see ScreenElement
     */
    public boolean showSelectionHandle() {
        return true;
    }

    /**
     * @return
     * @see ScreenElement
     */
    public boolean isDraggable() {
        return false;
    }

    /**
     * @see ScreenElement
     */
    protected boolean hasToolTipText() {
        return true;
    }

    /**
     * @see ScreenElement
     */
    protected String getToolTipText() {
        return String.valueOf(synapse.getToolTipText());
    }

    /**
     * @return
     * @see ScreenElement
     */
    public boolean hasContextMenu() {
        return true;
    }

    /**
     * @see ScreenElement
     */
    protected JPopupMenu getContextMenu() {

        // JPopupMenu contextMenu = new JPopupMenu();
        //
        // contextMenu.add(new CutAction(getNetworkPanel()));
        // contextMenu.add(new CopyAction(getNetworkPanel()));
        // contextMenu.add(new PasteAction(getNetworkPanel()));
        // contextMenu.addSeparator();
        //
        // contextMenu.add(new DeleteAction(getNetworkPanel()));
        // contextMenu.addSeparator();
        //
        // contextMenu.add(getNetworkPanel().getActionManager().getGroupMenu());
        // contextMenu.addSeparator();
        //
        // // Workspace workspace = getNetworkPanel().getWorkspace();
        // // if (workspace.getGaugeList().size() > 0) {
        // // contextMenu.add(workspace.getGaugeMenu(getNetworkPanel()));
        // // contextMenu.addSeparator();
        // // }
        //
        // contextMenu.add(new SetSynapsePropertiesAction(getNetworkPanel()));
        //
        // return contextMenu;
        return this.getNetworkPanel().getSynapseContextMenu(synapse);
    }

    /**
     * @see ScreenElement
     */
    protected boolean hasPropertyDialog() {
        return true;
    }

    /**
     * @see ScreenElement
     */
    protected JDialog getPropertyDialog() {
        SynapseDialog dialog = (SynapseDialog) getNetworkPanel().getSynapseDialog(getNetworkPanel().getSelectedSynapses());
        return dialog;
    }

    /**
     * Returns String representation of this NeuronNode.
     *
     * @return String representation of this node.
     */
    public String toString() {
        String ret = new String();
        ret += "SynapseNode: (" + this.getGlobalFullBounds().x + ")(" + getGlobalFullBounds().y + ")\n";
        return ret;
    }

    /**
     * @return Returns the synapse.
     */
    public Synapse getSynapse() {
        return synapse;
    }

    /**
     * @param synapse The synapse to set.
     */
    public void setSynapse(final Synapse synapse) {
        this.synapse = synapse;
    }

    /**
     * @return Returns the source.
     */
    public NeuronNode getSource() {
        return source;
    }

    /**
     * @return Returns the target.
     */
    public NeuronNode getTarget() {
        return target;
    }

    /**
     * @param source The source to set.
     */
    public void setSource(final NeuronNode source) {
        this.source = source;
    }

    /**
     * @param target The target to set.
     */
    public void setTarget(final NeuronNode target) {
        this.target = target;
    }

    /**
     * @see ScreenElement
     */
    public void resetColors() {
        line.setStrokePaint(lineColor);
        updateColor();
        updateDiameter();
    }

    /**
     * @return the excitatoryColor
     */
    public static Color getExcitatoryColor() {
        return excitatoryColor;
    }

    /**
     * @param excitatoryColor the excitatoryColor to set
     */
    public static void setExcitatoryColor(Color excitatoryColor) {
        SynapseNode.excitatoryColor = excitatoryColor;
    }

    /**
     * @return the inhibitoryColor
     */
    public static Color getInhibitoryColor() {
        return inhibitoryColor;
    }

    /**
     * @param inhibitoryColor the inhibitoryColor to set
     */
    public static void setInhibitoryColor(Color inhibitoryColor) {
        SynapseNode.inhibitoryColor = inhibitoryColor;
    }

    /**
     * @return the zeroWeightColor
     */
    public static Color getZeroWeightColor() {
        return zeroWeightColor;
    }

    /**
     * @param zeroWeightColor the zeroWeightColor to set
     */
    public static void setZeroWeightColor(Color zeroWeightColor) {
        SynapseNode.zeroWeightColor = zeroWeightColor;
    }

    /**
     * @return the maxDiameter
     */
    public static int getMaxDiameter() {
        return maxDiameter;
    }

    /**
     * @param maxDiameter the maxDiameter to set
     */
    public static void setMaxDiameter(int maxDiameter) {
        SynapseNode.maxDiameter = maxDiameter;
    }

    /**
     * @return the minDiameter
     */
    public static int getMinDiameter() {
        return minDiameter;
    }

    /**
     * @param minDiameter the minDiameter to set
     */
    public static void setMinDiameter(int minDiameter) {
        SynapseNode.minDiameter = minDiameter;
    }

    /**
     * @return the lineColor
     */
    public static Color getLineColor() {
        return lineColor;
    }

    /**
     * @param lineColor the lineColor to set
     */
    public static void setLineColor(Color lineColor) {
        SynapseNode.lineColor = lineColor;
    }

    /**
     * @return the line
     */
    public PPath.Float getLine() {
        return line;
    }

    public Arc2D.Float getArcBound() {
        return arcBound;
    }


    public Line2D.Float getLineBound() {
        return lineBound;
    }
}

File: src/main/kotlin/org/simbrain/network/gui/nodes/SynapseNode.kt
package org.simbrain.network.gui.nodes

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.swing.Swing
import org.piccolo2d.nodes.PPath
import org.piccolo2d.util.PBounds
import org.piccolo2d.util.PPaintContext
import org.simbrain.network.core.Synapse
import org.simbrain.network.gui.NetworkPanel
import org.simbrain.network.gui.createSynapseContextMenu
import org.simbrain.network.gui.createTooltipText
import org.simbrain.network.gui.dialogs.NetworkPreferences
import org.simbrain.network.gui.dialogs.NetworkPreferences.excitatorySynapseColor
import org.simbrain.network.gui.dialogs.NetworkPreferences.inhibitorySynapseColor
import org.simbrain.network.gui.dialogs.NetworkPreferences.maxWeightSize
import org.simbrain.network.gui.dialogs.NetworkPreferences.minWeightSize
import org.simbrain.network.gui.dialogs.NetworkPreferences.spikingColor
import org.simbrain.network.gui.dialogs.synapse.SynapseDialog
import org.simbrain.util.*
import java.awt.Color
import java.awt.geom.Arc2D
import java.awt.geom.Area
import java.awt.geom.Line2D
import java.awt.geom.Point2D
import javax.swing.JPopupMenu
import kotlin.math.*

/**
 *  Piccolo representation of [Synapse]
 */
class SynapseNode(
    net: NetworkPanel?,
    var source: NeuronNode,
    var target: NeuronNode,
    var synapse: Synapse
) : ScreenElement(net!!) {

    /**
     * Location of circle relative to target node.
     */
    private val offset = 7.0

    /**
     * Main circle of the synapse.
     */
    private var circle: PPath? = null

    /**
     * Line connecting nodes. More of a loop for self connections.
     */
    var line: Float? = null
        private set

    /**
     * The line for bound checking.
     */
    val lineBound: Line2D.Float = Line2D.Float()

    /**
     * The arc for bound checking when the synapse is a self connection.
     */
    var arcBound: Arc2D.Float = Arc2D.Float()
        private set

    init {
        updatePosition()
        this.addChild(circle)
        this.addChild(line)
        line!!.strokePaint = lineColor
        line!!.lowerToBottom()
        line!!.paint = null

        updateColor()
        updateDiameter()

        pickable = true
        circle!!.pickable = true
        line!!.pickable = false

        val events = synapse.events

        events.strengthUpdated.on(dispatcher = Dispatchers.Swing) {
            updateColor()
            updateDiameter()
        }
        events.colorPreferencesChanged.on(dispatcher = Dispatchers.Swing) {
            updateColor()
            updateDiameter()
            updateSpikeColor()
        }
        events.visbilityChanged.on(dispatcher = Dispatchers.Swing) { _, newVisibility ->
            updateVisibility(newVisibility) }
        updateVisibility(synapse.isVisible)
        events.clampChanged.on(dispatcher = Dispatchers.Swing) { this.updateClampStatus() }
        updateClampStatus()

        events.locationChanged.on(dispatcher = Dispatchers.Swing) { this.updatePosition() }

        // Respond to spiking events
        source.neuron.events.spiked.on(dispatcher = Dispatchers.Swing) {
            updateSpikeColor()
            // If spiking-only mode is on, adjust visibility in response to spike changes
            if (networkPanel.synapseSpikingOnlyVisible) {
                applySpikingOnlyVisibility()
            }
        }

        // Respond to global toggles for spiking-only visibility
        networkPanel.network.events.synapseSpikingOnlyVisibilityChanged.on(dispatcher = Dispatchers.Swing) {
            applySpikingOnlyVisibility()
        }
    }


    fun updatePosition() {
        // Position the synapse
        val synapseCenter = if (isSelfConnection) {
            globalToLocal(Point2D.Double(target.neuron.x + offset + 3, target.neuron.y + offset + 3))
        } else {
            globalToLocal(calcCenter(source.neuron.location, target.neuron.location))
        }
        this.offset(synapseCenter.x - offset, synapseCenter.y - offset)

        // Create the circle
        if (circle == null) {
            circle = createEllipse(0f, 0f, offset.toFloat() * 2, offset.toFloat() * 2)
            circle!!.strokePaint = null
        }
        val ea = 1.0
        val ea2 = ea * 2
        val tempBounds = circle!!.fullBounds
        setBounds(tempBounds.x - ea, tempBounds.y - ea, tempBounds.width + ea2, tempBounds.height + ea2)

        // Create the line
        if (line == null) {
            line = createLine(globalToLocal(synapseCenter))
        }

        // Update the line (unless it's a self connection)
        if (!isSelfConnection) {
            line!!.reset()
            line!!.append(Line2D.Double(globalToLocal(source.neuron.location), synapseCenter), false)
            lineBound.setLine(source.neuron.location, localToGlobal(synapseCenter))
        } else {
            arcBound =
                Arc2D.Float(
                    globalBounds.getX().toFloat(),
                    globalBounds.getY().toFloat() - 7,
                    22f,
                    15f,
                    1f,
                    355f,
                    Arc2D.OPEN
                )
        }
    }


    fun updateClampStatus() {
        if (synapse.clamped) {
            circle!!.strokePaint = NetworkTheme.current.nodeOutline
        } else {
            circle!!.strokePaint = null
        }
    }

    override fun refreshTheme() {
        super.refreshTheme()
        updateClampStatus()
    }

    val isSelfConnection: Boolean
        get() = (source.neuron == target.neuron)

    /**
     * Create the line depending on whether this is self connected or not.
     *
     * @param center the center of the synapse
     */
    private fun createLine(center: Point2D): Float {
        return if (isSelfConnection) {
            Float(Arc2D.Float(x.toFloat(), y.toFloat() - 7, 22f, 15f, 1f, 355f, Arc2D.OPEN))
        } else {
            Float(Line2D.Float(globalToLocal(source.center), center))
        }
    }

    /**
     * Calculates the color for a weight, based on its current strength.
     * Positive values are (for example) red, negative values blue.
     */
    fun updateColor() {
        if (synapse.strength < 0) {
            circle!!.paint = inhibitoryColor
        } else if (synapse.strength == 0.0) {
            circle!!.paint = zeroWeightColor
        } else {
            circle!!.paint = excitatoryColor
        }
    }

    /**
     * When spiking change the color of the line.
     */
    private fun updateSpikeColor() {
        if (with(networkPanel.network) { source.neuron.isSpike }) {
            line!!.strokePaint = spikingColor
        } else {
            line!!.strokePaint = lineColor
        }
    }

    /**
     * Apply combined logic for visibility:
     * - If the model's own visibility is false, this node must remain hidden.
     * - If spiking-only mode is enabled, only show while the source neuron is spiking.
     * - Otherwise reflect the model's own visibility.
     */
    private fun applySpikingOnlyVisibility() {
        // Old synapse visibility has higher priority
        if (!synapse.isVisible) {
            super.setVisible(false)
            return
        }
        if (networkPanel.synapseSpikingOnlyVisible) {
            val show = with(networkPanel.network) { source.neuron.isSpike }
            super.setVisible(show)
        } else {
            super.setVisible(true)
        }
    }

    private fun updateVisibility(modelVisibility: Boolean) {
        if (!modelVisibility) {
            super.setVisible(false)
            return
        }
        if (networkPanel.synapseSpikingOnlyVisible) {
            applySpikingOnlyVisibility()
        } else {
            super.setVisible(true)
        }
    }

    /**
     * Update the diameter of the drawn weight based on the logical weight's
     * strength.
     */
    fun updateDiameter() {
        val diameter: kotlin.Double

        var upperBound = synapse.upperBound
        var lowerBound = synapse.lowerBound
        var strength = synapse.strength

        // If upper or lower bound are set to zero use a proxy to prevent
        // division errors
        if (upperBound == 0.0) {
            upperBound = ZERO_PROXY
        }
        if (lowerBound == 0.0) {
            lowerBound = ZERO_PROXY
        }

        // If strength is out of bounds (which is allowed in the model), set it
        // to those bounds for the
        // sake of the GUI representation
        if (strength < lowerBound) {
            strength = lowerBound
        }
        if (strength > upperBound) {
            strength = upperBound
        }

        diameter = if (synapse.strength == 0.0) {
            minDiameter.toDouble()
        } else if (synapse.strength > 0) {
            (maxDiameter - minDiameter) * (strength / upperBound) + minDiameter
        } else {
            ((maxDiameter - minDiameter) * (abs(
                strength / lowerBound
            ))) + minDiameter
        }

        val delta = (circle!!.bounds.getWidth() - diameter) / 2

        circle!!.width = diameter
        circle!!.height = diameter
        // offset properly moves circle, but this is not reflected in bounds
        circle!!.offset(delta, delta)
        setBounds(circle!!.fullBounds)
    }

    /**
     * Calculates the position of the synapse circle based on the positions of
     * the source and target NeuronNodes.
     *
     * @param src Source NeuronNode
     * @param tar Target NeuronNode
     * @return the appropriate position for the synapse circle
     */
    fun calcCenter(src: Point2D, tar: Point2D): Point2D {
        val sourceX = src.x
        val sourceY = src.y
        val targetX = tar.x
        val targetY = tar.y

        if (sourceX == targetX && sourceY == targetY) {
            return Point2D.Double(0.0, 0.0)
        }

        val x = abs(sourceX - targetX)
        val y = abs(sourceY - targetY)
        val alpha = atan(y / x)

        var weightX = 0.0
        var weightY = 0.0

        val neuronOffset = NEURON_DIAMETER / 2

        weightX = if (sourceX < targetX) {
            targetX - (neuronOffset * cos(alpha))
        } else {
            targetX + (neuronOffset * cos(alpha))
        }

        weightY = if (sourceY < targetY) {
            targetY - (neuronOffset * sin(alpha))
        } else {
            targetY + (neuronOffset * sin(alpha))
        }

        return Point2D.Double(weightX, weightY)
    }

    override val isDraggable: Boolean
        get() = false


    protected fun hasToolTipText(): Boolean {
        return true
    }

    override val toolTipText: String
        get() = createTooltipText(synapse) {
            synapse.toString()
        }

    override val contextMenu: JPopupMenu
        get() = networkPanel.createSynapseContextMenu(synapse)

    override fun createEditDialog(): StandardDialog? {
        return SynapseDialog.createSynapseDialog(networkPanel.selectionManager.filterSelectedModels<Synapse>())
    }

    override val propertyDialog: StandardDialog? get() = createEditDialog()

    override fun toString(): String {
        var ret = ""
        ret += "SynapseNode: (" + this.globalFullBounds.x + ")(" + globalFullBounds.y + ")\n"
        return ret
    }

    override val model: Synapse
        get() = synapse


    override fun paintAfterChildren(paintContext: PPaintContext) {
        super.paintAfterChildren(paintContext)
        if (!NetworkPreferences.showSynapseStrengthLabels) return
        if (isSelfConnection) return

        val diameter = circle?.width ?: return
        if (diameter * networkPanel.scalingFactor < NetworkPreferences.synapseStrengthLabelMinScreenSize) return

        // Visible center of the circle in this node's local coords (fullBounds includes circle.offset).
        val centerRef = circle!!.fullBoundsReference
        val circleCx = centerRef.centerX
        val circleCy = centerRef.centerY

        // The synapse circle sits on the target neuron's perimeter, so the half toward the target is
        // covered. Offset the label toward the source (visible half). Direction is preserved under
        // translation, so we can compute it directly from the global neuron locations.
        val dx = source.neuron.x - target.neuron.x
        val dy = source.neuron.y - target.neuron.y
        val mag = sqrt(dx * dx + dy * dy)
        if (mag < 1e-6) return
        val labelX = circleCx + (dx / mag) * (diameter / 4)
        val labelY = circleCy + (dy / mag) * (diameter / 4)

        val decimals = NetworkPreferences.synapseStrengthDecimalPlaces
        val text = synapse.strength.format(decimals)
        val refString = "-9." + "9".repeat(decimals)
        val g2 = paintContext.graphics
        val font = computeCellFont(diameter * 0.6, diameter * 0.3, refString, g2.fontRenderContext)
        g2.drawCenteredOutlinedLabel(text, font, labelX, labelY)
    }

    override fun isIntersecting(bound: PBounds?): Boolean {
        if (isSelfConnection) {
            val boundArea = Area(bound)
            boundArea.intersect(Area((arcBound)))
            return !boundArea.isEmpty
        } else {
            return bound!!.intersectsLine(lineBound)
        }
    }

    companion object {
        /**
         * Used to approximate zero to prevent divide-by-zero errors.
         */
        private const val ZERO_PROXY = .001

        var excitatoryColor: Color = excitatorySynapseColor

        var inhibitoryColor: Color = inhibitorySynapseColor

        var zeroWeightColor: Color = NetworkPreferences.zeroWeightColor

        var maxDiameter: Int = maxWeightSize

        var minDiameter: Int = minWeightSize

        var lineColor: Color = NetworkPreferences.connectionLineColor
    }
}